### ISSUE TITLE
Queue triggered function takes long time to start

### ISSUE THREAD

**Issue by jmezach** [Azure/azure-functions-powershell-worker#490]
As discussed on [Twitter](https://twitter.com/jmezach/status/1283487533364633601) I'm seeing some weird behaviour with queue triggered functions written in PowerShell, or at least behaviour that I'm failing to understand. Looking at the logs it seems that the message from the queue is being read, but after that it takes quite some time before my function is actually going to run. This period of time also seems to grow over time. For example, I'm seeing this in my Logs:

```
2020-07-16 06:00:45.725
Executing 'Functions.RefreshNuGetAnalyticsForRepo' (Reason='New queue message detected on 'repositories'.', Id=f02f3719-1061-4aff-a643-10b9c41c3b21)
Information
2020-07-16 06:00:45.725
Trigger Details: MessageId: 891dbcdf-cb81-4f21-bc22-477cbf1152b7, DequeueCount: 1, InsertionTime: 2020-07-16T05:59:06.000+00:00
Information
2020-07-16 06:01:50.098
Your function 'RefreshNuGetAnalyticsForRepo' is queuing requests as there are no available runspaces. You may be able to increase your throughput by following the best practices on https://aka.ms/functions-powershell-concurrency.
Warning
2020-07-16 06:02:54.068
INFORMATION: Analyzing repository ToolingCluster...
Information
```

As you can see in this case it seems to take more than 2 minutes for the function to actually doing its work, but I've also seen cases where this takes more than 8 minutes. Usually these end up failing because of the 10 minute maximum execution time on the Consumption plan.

Obviously I've read through the link provided that talks about [concurrency](https://aka.ms/functions-powershell-concurrency) and tweaked the value of both `FUNCTIONS_WORKER_PROCESS_COUNT` and `PSWorkerInProcConcurrencyUpperBound`. Interestingly changing these parameters doesn't seem to have any effect on this behaviour. However, my script fails if I set both settings to a value higher than `1` due to some concurrency issues inherent in the script. But frankly, I don't mind, since this is a script that is going to run twice a day and it doesn't really matter to me if it takes a little bit longer.

Let me elaborate a little bit on what I'm trying to achieve. We've had this PowerShell script running as a scheduled task on an on-prem machine, which goes out to Azure DevOps, goes through each Git repository and fetches all the dependency information (NuGet packages) from them and drops that into a SQL Server database for analysis using Power BI. I'm currently trying to move that script and its database to the cloud, but I want to run this as cheaply as possible so I've rewritten the script to use Azure Table Storage as the storage. To run it I'm looking at Azure Functions. Initially I ran into a problem where the script was simply taking too long to process all repositories. That's why I came up with splitting the script into a smaller piece that processes a single repository that is triggered by a queue message.

I did just find an interesting bug in my script though, where it seems that the queue triggered function is processing files from another repository. I've now fixed that bug, so hopefully that will solve my current issue. But I still believe that this is an issue.

**Comment by AnatoliB** [comment]
@jmezach From our internal logs, both `FUNCTIONS_WORKER_PROCESS_COUNT` and `PSWorkerInProcConcurrencyUpperBound` seem to be at 1 between 2020-07-16 5:00 and 2020-07-16 6:30. If you believe you were actually trying to tweak them in this time interval, please let us know and we will focus on figuring this out.

Between 5:59 and 6:01, we detected 136 messages in the `repositories` queue. Each message resulted in a function `RefreshNuGetAnalyticsForRepo` invocation request. These requests were distributed between 10 instances, each of them having exactly one PowerShell worker process hosting exactly one PowerShell runspace (according to the settings above). In this configuration, each instance is forced to process just one request at a time. Processing each message took from a few seconds to ~2 minutes, and each instance had 8-14 messages to process. Please note that the 10 min timeout clock starts soon after the message is picked from the queue, _not_ when the processing actually starts. So, some requests timed out.

If you can resolve the concurrency issue in your scripts, increasing `PSWorkerInProcConcurrencyUpperBound` or `FUNCTIONS_WORKER_PROCESS_COUNT` would be the easiest solution. Your CPU consumption seems to be low, so increasing concurrency should increase the processing rate and get rid of timeouts. Please let us know if you need any help with that.

**Comment by AnatoliB** [comment]
@jmezach Also, if performance is not your concern but you just need to reliably process all the queued messages, consider triggering your function on a timer, so that your function code can query the queue and process messages as appropriate, perhaps even one at a time if required.

**Comment by jmezach** [comment]
Since I've fixed the issue in my script I see that my solution works as expected. I had a bug where the time needed to process each message in the queue was getting longer exponentially because some state was being shared between different executions of the function since I was writing some files to disk. Since I've fixed that I'm no longer seeing timeouts for the queue triggered function.

I did notice a timeout for my timer triggered function which runs first and then puts a message into the queue for each repository to process. I'm guessing that was caused by some slowness on the Azure DevOps side. I'll try to add some more diagnostics there so I can better see what's going on, but I think it is unrelated to this issue. As far as I'm concerned I think we can close this issue.
